### PR TITLE
Add boundary data overlay to evaluation plots

### LINKED ISSUE #108: Plotting Boundaries in `train_model.py --eval` not just interior
With milestone v0.4.0 we are releasing realistic boundaries. For that purpose the plotting functionalities need to be updated. Currently they only plot the interior domain, with a semi-transparent boundary around it. In the new approach the boundary and interior data comes from two separate datastores. 

Some considerations:

- Using domain_cropping in the config will keep only the points at the given distance from the convex hull of the referenced interior data. This set of points will in general not lay on a regular grid.
- There is no boundary mask in neural-lam right now. All points in the boundary datastore are considered boundary input points.
- As plots anyhow are in the crs of the interior, boundary points would still not lay on a regular grid (if they were originally in e.g. a regular lat-lon grid).

I started working on this but got lost in the x,y,lat,lon,projections,coordinates. When are they 1D, when are they 2D, stacked unstacked. Anyways :D here is my working branch: https://github.com/sadamov/neural-lam/tree/plot_boundaries But unfortunately I don't have the bandwidth right now to continue working on it. A typical example plot looks like this:

Image

### PR DESCRIPTION
**Pull request by sadamov**
## Describe your changes

Wire the boundary datastore from `config.yaml` into `ForecasterModule` and the plotting pipeline so that evaluation plots (`--eval`) overlay boundary forcing data from a separate domain underneath the interior prediction and ground truth panels.

Changes:
- `vis.plot_on_axis` and `vis.plot_prediction` accept optional `boundary_da` and `boundary_datastore` parameters. When provided, boundary data is plotted underneath the interior field at alpha=0.5 and the map extent is expanded by a configurable `boundary_margin_degrees` margin (default 1°) around the interior domain.
- Boundary grid coordinates are read via `get_xy("forcing")` / `get_lat_lon("forcing")` so that boundary datastores without state variables (e.g. ERA5 forcing-only) work correctly.
- New `plotting` section in `config.yaml` (`NeuralLAMConfig.plotting`, a `PlottingConfig`): `boundary_datastore` names which entry in `datastores` supplies the overlay (defaults to the sole boundary datastore), `boundary_margin_degrees` controls the map extent margin, and `boundary_var_mapping` maps interior state variable names to boundary forcing feature names (falling back to same-name matching).
- `ForecasterModule.__init__` accepts an optional `datastore_boundary`. During `plot_examples`, raw (unstandardized) boundary forcing is loaded from the boundary datastore and passed through to `vis.plot_prediction`, matched per state variable via `boundary_var_mapping`.
- `train_model.py` passes `datastore_boundary` from config into `ForecasterModule`.
- Deprecate (rather than remove) the `crop_to_interior` keyword on `plot_prediction`/`plot_spatial_error`, emitting a warning and ignoring it, to avoid a breaking change for existing callers.
- Integration test with real DANRA + ERA5 boundary data (marked `@pytest.mark.slow`); fixed `mllam-data-prep` resolving a referenced datastore config relative to the process CWD instead of the config file's directory.
- README updated to document the boundary overlay behavior and the new `plotting` config options during evaluation.

**Depends on #635** (boundary datastore data loading). Please merge #635 first; after that the base branch of this PR can be updated to `main`.

## Issue Link

refs #108

## Type of change

- [ ] 🐛 Bug fix (non-breaking change that fixes an issue)
- [x] ✨ New feature (non-breaking change that adds functionality)
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] 📖 Documentation (Addition or improvements to documentation)

## Checklist before requesting a review

- [x] My branch is up-to-date with the target branch - if not update your fork with the changes from the target branch (use `pull` with `--rebase` option if possible).
- [x] I have performed a self-review of my code
- [x] For any new/modified functions/classes I have added docstrings that clearly describe its purpose, expected inputs and returned values
- [x] I have placed in-line comments to clarify the intent of any hard-to-understand passages of my code
- [x] I have updated the [README](README.MD) to cover introduced code changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have given the PR a name that clearly describes the change, written in imperative form ([context](https://www.gitkraken.com/learn/git/best-practices/git-commit-message#using-imperative-verb-form)).
- [x] I have requested a reviewer and an assignee (assignee is responsible for merging). This applies only if you have write access to the repo, otherwise feel free to tag a maintainer to add a reviewer and assignee.

## Checklist for reviewers

Each PR comes with its own improvements and flaws. The reviewer should check the following:
- [ ] the code is readable
- [ ] the code is well tested
- [ ] the code is documented (including return types and parameters)
- [ ] the code is easy to maintain

## Author checklist after completed review

- [x] I have added a line to the CHANGELOG describing this change, in a section
  reflecting type of change (add section where missing):
  - *added*: when you have added new functionality
  - *changed*: when default behaviour of the code has been changed
  - *fixes*: when your contribution fixes a bug
  - *maintenance*: when your contribution is relates to repo maintenance, e.g. CI/CD or documentation

## Checklist for assignee

- [ ] PR is up to date with the base branch
- [ ] the tests pass
- [ ] (if the PR is not just maintenance/bugfix) the PR is assigned to the next milestone. If it is not, propose it for a future milestone.
- [ ] author has added an entry to the changelog (and designated the change as *added*, *changed*, *fixed* or *maintenance*)
- Once the PR is ready to be merged, squash commits and merge the PR.

closes #108

### CODE REVIEW

**Comment by sadamov** [comment]
here are some examples for dummy data with and without boundary each:

boundary_plot_interior_only
boundary_plot_with_overlay

File: neural_lam/vis.py [hunk 1]
```diff
@@ -240,6 +287,8 @@ def plot_prediction(
     boundary_alpha=0.7,
     crop_to_interior=True,
```
Comment (kshirajahere) [inline neural_lam/vis.py]: the default eval path still crops to the interior. plot_examples() calls plot_prediction() without overriding crop_to_interior, so boundary data is drawn and then set_extent crops back to the interior; with a 1000 km boundary dummy, I saw the called extent stay at [5.755, 14.251, 53.693, 58.245] while the boundary extent is [0.964, 19.036, 51.277, 60.49]. When boundary_da is provided, either default to the boundary/union extent or explicitly pass crop_to_interior=False from plot_examples().

File: neural_lam/vis.py [hunk 2]
```diff
@@ -84,6 +93,37 @@ def plot_on_axis(
     gl.xlabel_style = {"size": _TICK_SIZE}
     gl.ylabel_style = {"size": _TICK_SIZE}
 
+    # Plot boundary data underneath interior data when provided
+    if boundary_da is not None and boundary_datastore is not None:
+        b_lats_lons = boundary_datastore.get_lat_lon("state")
```
Comment (kshirajahere) [inline neural_lam/vis.py]: this uses boundary_datastore.get_lat_lon("state"), but the PR description says boundary-only datastores without state variables are supported via forcing coordinates. Current #636 head also lacks #635’s later boundary-only MDP fixes. Use the forcing category here and add a test where the boundary datastore genuinely has no state

File: neural_lam/models/module.py [hunk 3]
```diff
@@ -443,6 +465,19 @@ def plot_examples(self, batch, n_examples, split, prediction):
...
+                # Select boundary field for this timestep (if available)
+                boundary_da_t = None
+                if (
+                    da_boundary_forcing is not None
+                    and "time" in da_boundary_forcing.dims
+                ):
+                    target_time = np.array(
+                        time_slice[t_i - 1].cpu(), dtype="datetime64[ns]"
+                    )
+                    boundary_da_t = da_boundary_forcing.sel(
```
Comment (kshirajahere) [inline neural_lam/models/module.py:782]: method="nearest" can silently overlay the wrong boundary timestep when the boundary datastore has missing times or a different cadence. For verification plots this should be exact, or at least use an explicit tolerance tied to the datastore step length and fail/skip when no valid boundary time exists.

**Comment by sadamov** [comment]
thanks @kshirajahere for the review, I'll have a look soon.
In the meantime I also implemented a "real-world" test-case using DANRA and ERA5 data across #635 and #636. The resulting plot looks like this: 
image

PS: Tests fail also on main right now because of missing torch wheel for 2.12.0.

**Comment by kshirajahere** [comment]
The plots look amazing @sadamov :D

**Comment by joeloskarsson** [comment]
I see the extent of the plot is now determined based on the full boundary size (please clarify if I missunderstood). I wonder if we want to give the user some more options here, for example if the boundary is very large one might want to only zoom in on the interior, with this covering most of the plot. Not sure where such an option best sits though.

**Comment by sadamov** [comment]
> I see the extent of the plot is now determined based on the full boundary size (please clarify if I missunderstood). I wonder if we want to give the user some more options here, for example if the boundary is very large one might want to only zoom in on the interior, with this covering most of the plot. Not sure where such an option best sits though.

not quite. first ,there is a simple check (boundary data present yes/no), if no -> plot interior only. If yes -> then plot a 1° margin (in projection) around the interior. this 1° is independent of the full boundary size and can be set via `boundary_margin_degrees`. I did however not expose this to the user yet and it would have to be wired through ForecasterModule.plot_examples, if you wanted that.

File: neural_lam/vis.py (hunk 1 above)
Comment (sadamov) [reply neural_lam/vis.py]: i changed the logic here to a clearly defined margin in degrees (projection) around the interior box. I hope that is clearer to use

File: neural_lam/vis.py (hunk 2 above)
Comment (sadamov) [reply neural_lam/vis.py]: you are right! I did a bit of cheating with the inheritance of DummyBoundaryDatastore. Now I properly removed the state (3e6e246, part of #635) and added a plotting test for the no-state path (a1789b2).

Comment (sadamov) [reply neural_lam/vis.py]: oh yeah and I also added the full era5 reanalysis dataset for testing with only forcing vars selected in the yaml config for mllam-data-prep. so now we should be fully covered on the forcing-only aspect

File: neural_lam/models/module.py (hunk 3 above)
Comment (sadamov) [reply neural_lam/models/module.py:782]: I actually like this to be nearest. Often the boundary will have coarser temporal resolution (e.g. 6h) and than the interior e.g. 1h. Now, neural-lam allows to select any number of future and past timesteps from the buondary that are nearest to that interior timestep creating a window. And I think it makes sense to display the nearest boundary in the verification plot because that is the boundary forcing closest to the truth that neural-lam can use in the inference. So unless you have strong objections, i'd rather leave this as is.

Comment (sadamov) [reply neural_lam/models/module.py:782]: I added a comment in the code about this rationale: 7b82b07

**Comment by joeloskarsson** [comment]
Aha, I see 😄 Yea I guess wiring through`boundary_margin_degrees` to the user is exactly what I was thinking about then. To follow all other plotting config, this should be an argparse flag. But this also feels like we are getting a ridiculous amount of flags 😮‍💨 Not sure if it is worth it. Maybe an option is to group this with the configuration of interior-boundary variable-mappings somehow? That is also plotting config really.

**Comment by sadamov** [comment]
> Aha, I see 😄 Yea I guess wiring through`boundary_margin_degrees` to the user is exactly what I was thinking about then. To follow all other plotting config, this should be an argparse flag. But this also feels like we are getting a ridiculous amount of flags 😮‍💨 Not sure if it is worth it. Maybe an option is to group this with the configuration of interior-boundary variable-mappings somehow? That is also plotting config really.

I introduce a new plotting config here: https://github.com/mllam/neural-lam/pull/636/commits/d6acd0f992cf8a9b3b4d4a2d37cda1ef8e1df957 for both vars matching (similar to research branch I think) and boundary width.

File: tests/datastore_examples/mdp/era5_1000hPa_danra_100m_winds/era5.datastore.yaml [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+        start: 2022-03-30T00:00
+        end: 2022-04-12T00:00
+  # Second-pass crop: keep only ERA5 points within `margin_width_degrees`
+  # of the interior dataset's convex hull. Follows the documented mllam
+  # pattern in `example.era5_cropped.yaml` -- the dim_mapping below stacks
+  # `[longitude, latitude]` directly into `grid_index` (no rename) so that
+  # convex-hull cropping can find lat/lon coords by name.
+  domain_cropping:
+    margin_width_degrees: 10
+    interior_dataset_config_path: danra.datastore.yaml
```
Comment (kshirajahere) [inline tests/datastore_examples/mdp/era5_1000hPa_danra_100m_winds/era5.datastore.yaml:37]: the slow “real-world” plotting test currently fails before it reaches plotting because `mllam-data-prep` resolves this as `danra.datastore.yaml` from the process cwd, not from this config file’s directory. Running `python -m pytest tests/test_plotting.py::test_plot_prediction_with_era5_boundary --run-slow -q` from the repo root raises `FileNotFoundError: [Errno 2] No such file or directory: 'danra.datastore.yaml'`, so the integration test cited for DANRA + ERA5 is not actually runnable.

File: neural_lam/vis.py [hunk 5]
```diff
@@ -483,12 +635,19 @@ def plot_prediction(
     title=None,
     vrange=None,
     boundary_alpha=0.7,
-    crop_to_interior=True,
     colorbar_label: str = "",
+    boundary_da=None,
+    boundary_datastore=None,
+    boundary_margin_degrees: float = 1.0,
```
Comment (kshirajahere) [inline neural_lam/vis.py:769]: removing the existing `crop_to_interior` keyword from `plot_prediction`/`plot_spatial_error` is a breaking API change for user scripts. On current `main`, callers can pass `crop_to_interior=False`; on this branch that raises `TypeError: got an unexpected keyword argument 'crop_to_interior'`.

File: README.md [hunk 6]
```diff
@@ -514,6 +526,23 @@ Some options specifically important for evaluation are:
 * `--n_example_pred`: Number of example predictions to plot during evaluation.
 * `--ar_steps_eval`: Number of time steps to unroll for during evaluation
 
+When a `datastore_boundary` is configured in `config.yaml`, evaluation plots
+will overlay the boundary forcing data underneath the interior prediction and
+ground truth panels, allowing visual inspection of the boundary conditions used
```
Comment (kshirajahere) [inline README.md]: this says the plotted boundary conditions are “used during the forecast”, but the current model path still ignores `boundary_features` in `ForecasterModule.common_step`; this PR loads boundary forcing separately for visualization in `plot_examples`.

**Review by kshirajahere — CHANGES_REQUESTED** [review]
Fascinating work. Thanks for giving me the opportunity to review it @sadamov really grateful :) left some 3 comments which I think might be worth addressing

File: tests/datastore_examples/mdp/era5_1000hPa_danra_100m_winds/era5.datastore.yaml (hunk 4 above)
Comment (sadamov) [reply tests/datastore_examples/mdp/era5_1000hPa_danra_100m_winds/era5.datastore.yaml:37]: ah true I didn't consider how mdp determines the output path. now it's fixed and documented in some comments

File: neural_lam/vis.py (hunk 5 above)
Comment (sadamov) [reply neural_lam/vis.py:769]: ok, it's probably a bad idea to remove this. I added a deprecation warning instead.

File: README.md (hunk 6 above)
Comment (sadamov) [reply README.md]: yes, here I was foreshadowing the upcoming PR that introduces boundary handling on the model side. I changed the comment to reflect the current state and this PR only.

**Comment by sadamov** [comment]
This is now rebased onto #635 and is ready for a review, too. @joeloskarsson
